File: 프로그래머스/2/72412. 순위 검색/순위 검색.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

unordered_map<string, vector<int>> m;

void addCase(vector<string> v, int score) {
    for(int i=0; i<16; i++) {
        int num = i;
        string tmp = "";
        for(int j=0; j<4; j++) {
            if(num<=0 || num %2 == 0) tmp += '-';
            else tmp += v[j];
            num /= 2;
        }
        m[tmp].push_back(score);
    }
}

vector<int> solution(vector<string> info, vector<string> query) {
    vector <string> v(5);
    vector <int> ans;
    for(auto i : info) {
        istringstream ss(i);
        ss >> v[0] >> v[1] >> v[2] >> v[3] >> v[4];
        addCase(v, stoi(v[4]));
    }
    
    for(auto iter = m.begin(); iter!=m.end(); iter++) {
        sort(iter->second.begin(), iter->second.end());
    }
    
    for(auto q : query) {
        istringstream ss(q);
        int score;
        ss >> v[0] >> v[4] >> v[1] >> v[4] >> v[2] >> v[4] >> v[3] >> score;
        
        vector<int> v2 = m[v[0]+v[1]+v[2]+v[3]];
        if(v2.empty()) {
            ans.push_back(0);
            continue;
        }
        
        int k = lower_bound(v2.begin(), v2.end(), score) - v2.begin();
        ans.push_back(v2.size() - k);
    }
    return ans;
}

File: 프로그래머스/2/72412. 순위 검색/순위 검색.java
import java.util.*;
import java.io.*;

class Solution {
    Map<String, List<Integer>> map = new HashMap<>();
    
    public int[] solution(String[] info, String[] query) {
        for(String i : info) {
            String[] each = i.split(" ");
            List<String> combarr = new ArrayList<>();
            
            for(String e : each) {
                combarr.add(e);
            }
            combination(0, "", combarr);
        }
        
        for(List<Integer> m : map.values()) {
            Collections.sort(m);
        }
        
        List<Integer> result = new ArrayList<>();
        
        for(String q : query) {
            String[] splits = q.split(" ");
            String key = splits[0] + " and " + splits[2] + " and " + splits[4] + " and " + splits[6];
            int score = Integer.parseInt(splits[7]);
            
            if(!map.containsKey(key)) {
                result.add(0);
                continue;
            }

            List<Integer> eachmap = map.get(key);
            int eachresult = lowerbound(score, eachmap);
            result.add(eachresult);
        }
        return result.stream().mapToInt(Integer::intValue).toArray();
    }
    
    public int lowerbound(int score, List<Integer> eachmap) {
        int left = 0;
        int right = eachmap.size();
        
        while(left < right) {
            int mid = (left + right) / 2;
            
            if(eachmap.get(mid) >= score) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return eachmap.size() - left;
    }
    
    public void combination(int cnt, String result, List<String> combarr) {
        if(cnt == 4) {
            int score = Integer.parseInt(combarr.get(4));
            map.computeIfAbsent(result, k -> new ArrayList<>())
                .add(score);
            return;
        }
        
        if(cnt >= 1) {
            result += " and ";
        }
        
        combination(cnt+1, result + combarr.get(cnt), combarr);
        combination(cnt+1, result + "-", combarr);
    }
}